src/kvstore/LogEncoder.cpp

```diff
@@ -164,6 +164,61 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
     return values;
 }
 
+std::string encodeBatchValue(const std::vector<std::pair<BatchLogType,
+                             std::pair<std::string, std::string>>>& batch) {
+    auto type = LogType::OP_BATCH_WRITE;
+    std::string encoded;
+    encoded.reserve(1024);
+
+    // Timestamp (8 bytes)
+    int64_t ts = time::WallClock::fastNowInMilliSec();
+    encoded.append(reinterpret_cast<char*>(&ts), sizeof(int64_t));
+    // Log type
+    encoded.append(reinterpret_cast<char*>(&type), 1);
+    // Number of values
+    auto num = static_cast<uint32_t>(batch.size());
+    encoded.append(reinterpret_cast<char*>(&num), sizeof(uint32_t));
+    // Values
+    for (auto& op : batch) {
+        auto bType = op.first;
+        auto bPair = op.second;
+        encoded.append(reinterpret_cast<char*>(&bType), 1);
+        auto v1 = bPair.first;
+        auto v2 = bPair.second;
+        auto len = v1.size();
+        encoded.append(reinterpret_cast<char*>(&len), sizeof(uint32_t));
+        encoded.append(v1.c_str(), len);
+        len = v2.size();
+        encoded.append(reinterpret_cast<char*>(&len), sizeof(uint32_t));
+        encoded.append(v2.c_str(), len);
+    }
+
+    return encoded;
+}
+
+std::vector<std::pair<BatchLogType, std::pair<std::string, std::string>>>
```

Comment:
Could we return std::pair<folly::stringPiece,  folly::stringPiece>

Reply:
Wow, I had an epiphany that original return type will be memory copy occur.
Good cache. and good prompt.